File: 100joon/Sliver/2343.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())
lecture = list(map(int, input().split()))

answer = max(lecture)
left = answer
right = sum(lecture)
while left <= right:
    mid = (left + right) // 2
    cnt = 1
    size = 0
    for i in lecture:
        if size + i > mid:
            size = i
            cnt += 1
        else:
            size += i

    if cnt <= m:
        answer = mid
        right = mid - 1
    else:
        left = mid + 1

print(answer)

File: 100joon/Sliver/2343.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [n, m] = input[0].split(" ").map(Number);
const lectures = input[1].split(" ").map(Number);
let ans = Math.max(...lectures);
let [left, right] = [ans, lectures.reduce((acc, curr) => acc + curr)];

while (left <= right) {
  const mid = Math.floor((left + right) / 2);
  let sum = 0;
  let cnt = 1;

  for (let lecture of lectures) {
    if (sum + lecture > mid) {
      sum = lecture;
      cnt += 1;
    } else {
      sum += lecture;
    }
  }

  if (cnt <= m) {
    ans = mid;
    right = mid - 1;
  } else {
    left = mid + 1;
  }
}
console.log(ans);
